src/scripts/fetchStatsFromGA.js

```diff
@@ -0,0 +1,437 @@
+  let results = {};
+  let hasMore = false;
+  requestDocTypes.forEach((docType, i) => {
+    let report = reports[i];
+    console.log(
+      `fetched ${report.data.rows.length} starting at ${pageTokens[docType] ||
+        0} ` +
+        `out of total ${
+          report.data.rowCount
+        } rows for ${sourceType} ${docType}` +
+        ` with next pageToken ${report.nextPageToken}`
+    );
+    results[docType] = report.data.rows;
+    nextPageTokens[docType] = report.nextPageToken || -1;
+    hasMore = hasMore || report.nextPageToken != null;
+  });
+
+  return { results, pageTokens: nextPageTokens, hasMore };
+};
+
+const upsertDocStats = async function(body) {
+  try {
+    const res = await client.bulk({
+      body,
+      refresh: 'true',
+    });
+    if (res.body.errors) {
+      console.error(
+        `bulk upsert failed : ${JSON.stringify(res.body, null, '  ')}`
+      );
+    }
+  } catch (err) {
+    console.error(err);
+  }
+};
+
+async function fetchReplyUsers(rows) {
+  const replyIds = rows.map(row => parseIdFromRow(row));
+  const {
+    body: {
+      hits: { hits: replies },
+    },
+  } = await client.search({
+    index: 'replies',
+    size: pageSize,
+    body: {
+      query: {
+        ids: { values: replyIds },
+      },
+      _source: {
+        includes: ['userId'],
+      },
+    },
+  });
+  let replyUsers = {};
+  replies.forEach(reply => (replyUsers[reply._id] = reply._source.userId));
+  return replyUsers;
+}
+
+/**
+ * Given stats report for a sourceType and docType, store them in db
+ *
+ * @param {string} sourceType    One of the values defined by statsSources.
+ * @param {string} docType       One of the values defined by docTypes.
+ * @param {array}  rows          Rows of data to process; each row is an object
+    of the form {dimensions: [{field: value}], metrics:[{values:[values]}]}}
+ * @param {Date}   fetchedAt     Timestamp of the fetched time
+ * @param {object} prevLastRow   Stats for the last processed row from previous
+    call of the same sourceType and docType; It's an object of the form {
+      sourceType: string, docType: string, path: string, date: string,
+      docId: string, visits: number, users: number }
+
+ * @return {object} Return the stats of last processed row, same format as prevLastRow.
+ */
+const processReport = async function(
+  sourceType,
+  docType,
+  rows,
+  fetchedAt,
+  prevLastRow
+) {
+  let replyUsers;
+  if (docType === docTypes.REPLY) {
+    replyUsers = await fetchReplyUsers(rows);
+  }
+
+  let bulkUpdates = [];
+  const sourceName = sourceType.toLowerCase();
+
+  rows.forEach(row => {
+    const docId = parseIdFromRow(row);
+    const date = formatDate(row.dimensions[1]);
+    const [visits, users] = row.metrics[0].values;
+    let docUserId;
+
+    const prevParams =
+      bulkUpdates.length > 0 &&
+      bulkUpdates[bulkUpdates.length - 1].script.params;
+    if (prevParams && prevParams.date === date && prevParams.docId === docId) {
```

Comment:
I find it difficult to differentiate the use case between L291-294 and L307-315.

A comment elaborating this case (edge-case handling across batches) would be helpful.

Reply:
Another idea that come across my mind is, processing items across batches should be pretty similar to merging different `pagePathLevel2` -- they all memorize the last row and perform update when necessary.

This means that in theory we can just implement merging of different `pagePathLevel2`. When batch changes, we should put `prevLastRow` into the "last-row" memory and perform identical logic that handles `pagePathLevel2`.

Another way of thinking is that we blur the boundary of "batch" and only work on one row at a time, providing last "row" (no matter it's from same batch or previous batch) during the process.